File: client_side/client_app/lib/models/service_request.dart
class ServiceRequest {
  final String id;
  final String? userID;
  final String category;
  final String customerName;
  final String phone;
  final String address;
  final String? description;
  final DateTime preferredDate;
  final String preferredTime;
  final String
      status; // pending | approved | in-progress | completed | cancelled
  final String? assigneeId;
  final String? assigneeName;
  final String? assigneePhone;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  ServiceRequest({
    required this.id,
    this.userID,
    required this.category,
    required this.customerName,
    required this.phone,
    required this.address,
    this.description,
    required this.preferredDate,
    required this.preferredTime,
    required this.status,
    this.assigneeId,
    this.assigneeName,
    this.assigneePhone,
    this.createdAt,
    this.updatedAt,
  });

  factory ServiceRequest.fromJson(Map<String, dynamic> json) => ServiceRequest(
        id: (json['id'] ?? json['_id']).toString(),
        userID: json['userID']?.toString(),
        category: json['category']?.toString() ?? '',
        customerName: json['customerName']?.toString() ?? '',
        phone: json['phone']?.toString() ?? '',
        address: json['address']?.toString() ?? '',
        description: json['description']?.toString(),
        preferredDate:
            DateTime.tryParse(json['preferredDate']?.toString() ?? '') ??
                DateTime.now(),
        preferredTime: json['preferredTime']?.toString() ?? '',
        status: json['status']?.toString() ?? 'pending',
        assigneeId: json['assigneeId']?.toString(),
        assigneeName: json['assigneeName']?.toString(),
        assigneePhone: json['assigneePhone']?.toString(),
        createdAt: json['createdAt'] != null
            ? DateTime.tryParse(json['createdAt'].toString())
            : null,
        updatedAt: json['updatedAt'] != null
            ? DateTime.tryParse(json['updatedAt'].toString())
            : null,
      );
}

File: server_side/online_store_api/model/serviceRequest.js
const mongoose = require('mongoose');

const SERVICE_STATUS = ['pending', 'approved', 'in-progress', 'completed', 'cancelled'];

const serviceRequestSchema = new mongoose.Schema({
    userID: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    category: { type: String, required: true, trim: true },
    customerName: { type: String, required: true, trim: true },
    phone: { type: String, required: true, trim: true },
    address: { type: String, required: true, trim: true },
    description: { type: String, trim: true },
    preferredDate: { type: Date, required: true },
    preferredTime: { type: String, required: true },
    status: { type: String, enum: SERVICE_STATUS, default: 'pending', index: true },
    assigneeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Technician' },
    assigneeName: { type: String, trim: true },
    assigneePhone: { type: String, trim: true },
    notes: { type: String, trim: true }
}, { timestamps: true });

// Helpful indexes for common queries
serviceRequestSchema.index({ userID: 1, createdAt: -1 });
serviceRequestSchema.index({ phone: 1, createdAt: -1 });
serviceRequestSchema.index({ status: 1, createdAt: -1 });

const ServiceRequest = mongoose.model('ServiceRequest', serviceRequestSchema);

module.exports = { ServiceRequest, SERVICE_STATUS };
